packages/next/next-server/server/next-server.ts

```diff
@@ -324,10 +325,32 @@ export default class Server {
 
       let defaultLocale = i18n.defaultLocale
       let detectedLocale = detectLocaleCookie(req, i18n.locales)
-      let acceptPreferredLocale =
-        i18n.localeDetection !== false
-          ? accept.language(req.headers['accept-language'], i18n.locales)
-          : detectedLocale
+      let languages = new Negotiator(req).languages()
+      let acceptPreferredLocale = detectedLocale
+
+      if (i18n.localeDetection !== false) {
+        try {
+          acceptPreferredLocale = ResolveLocale(
+            new Set(i18n.locales),
+            languages,
+            {
+              localeMatcher: 'best fit',
+            },
+            [],
+            {},
+            () => defaultLocale
+          ).locale
+        } catch (e) {
+          // TODO: Log this or bubble this up, locale can be malformed
+        }
+
+        if (
+          acceptPreferredLocale === defaultLocale &&
+          languages[0] !== defaultLocale
+        ) {
+          // TODO: Bubble this up since we're serving default locale instead of preferred
```

Comment:
This piece if analytics is fairly important but not sure how this is done in next.js?

Reply:
Crty 2345 dfenga try